### ISSUE TITLE
Input Sanitizer missing characters which may appear in email addresses

### ISSUE THREAD

**Issue by knisbet** [gravitational/teleport#2025]
Version: Teleport Enterprise v2.6.2 git:v2.6.2-0-g2fa98e5f

Hey, so I ran into an issue with the input sanitizer being a little unhappy with my email address when using oidc (Auth0). I'm using the gmail auto tagging feature, where I use email a+tag@gmail.com, which will automatically assign a label to my email address. However, it looks like when this generates a dynamic identity from the oidc claims, and then runs the sanitizer, it doesn't like the username. 

```
Jun 20 04:30:27 ca-node-0 teleport[5803]: WARN             [OIDC] Error while processing callback: special characters are not allowed in resource names, please use name composed only from characters, hyphens and dots web/apiserver.go:814
```

Looking at the regex, it looks like it might be missing several valid email characters:
https://github.com/gravitational/teleport/blob/master/lib/backend/sanitize.go#L19

**Comment by benarent** [comment]
+1 this. I ran into this today.

SSO user login failed [special characters are not allowed in resource names, please use name composed only from characters, hyphens and dots]

**Comment by benarent** [comment]
I ran into this again with Teleport Cloud

**Comment by alex-kovoy** [comment]
We fixed it on the cloud side by not allowing the `+` emails

**Comment by knisbet** [comment]
I don't believe preventing valid emails on the cloud side fixes the issue in teleport @alex-kovoy ?

**Comment by programmerq** [comment]
I also ran into this recently. The teleport resource name regex that causes this is https://github.com/gravitational/teleport/blob/3149d0b9531fa178570239681b278c379a36cbb5/lib/backend/sanitize.go#L35

[RFC 5322](https://datatracker.ietf.org/doc/html/rfc5322#section-3.4) is much much more permissive for what is allowed in an email address. Email address matching can get quite complicated-- here's a regex that attempts to match any 5322 compliant address:

```
\A(?:[a-z0-9!#$%&'*+/=?^_‘{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_‘{|}~-]+)*
 |  "(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]
      |  \\[\x01-\x09\x0b\x0c\x0e-\x7f])*")
@ (?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?
  |  \[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}
       (?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:
          (?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]
          |  \\[\x01-\x09\x0b\x0c\x0e-\x7f])+)
     \])\z
```

While it is true that our resource sanitizer name validator shouldn't be checking for email address validity, it probably should allow many more of the characters that do appear in email addresses.

**Comment by knisbet** [comment]
> I also ran into this recently. The teleport resource name regex that causes this is
> 
> https://github.com/gravitational/teleport/blob/3149d0b9531fa178570239681b278c379a36cbb5/lib/backend/sanitize.go#L35
> 
> [RFC 5322](https://datatracker.ietf.org/doc/html/rfc5322#section-3.4) is much much more permissive for what is allowed in an email address. Email address matching can get quite complicated-- here's a regex that attempts to match any 5322 compliant address:
> 
> ```
> \A(?:[a-z0-9!#$%&'*+/=?^_‘{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_‘{|}~-]+)*
>  |  "(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]
>       |  \\[\x01-\x09\x0b\x0c\x0e-\x7f])*")
> @ (?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?
>   |  \[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}
>        (?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:
>           (?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]
>           |  \\[\x01-\x09\x0b\x0c\x0e-\x7f])+)
>      \])\z
> ```
> 
> While it is true that our resource sanitizer name validator shouldn't be checking for email address validity, it probably should allow many more of the characters that do appear in email addresses.

Just note, this whitelist pattern isn't to actually validate the email address, but to prevent problems with characters getting passed to the filesystem that can allow directory traversals and things like that. So RFC 5322 Regex validation may not also meet this requirement.

**Comment by wadells** [comment]
Duplicate of https://github.com/gravitational/teleport/issues/6632.  Fixed by https://github.com/gravitational/teleport/pull/8396.
